File: app/src/main/java/avanger/co/id/MainActivity.java
package avanger.co.id;

import androidx.appcompat.app.AppCompatActivity;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;

import android.Manifest;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.content.pm.PackageManager;
import android.os.Bundle;
import android.widget.Button;
import android.widget.EditText;
import android.widget.Toast;

public class MainActivity extends AppCompatActivity {
    private static final int cameraRequestCode = 1;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        // Pengaturan Shared Preferences.
        SharedPreferences sharedPref = getApplicationContext().getSharedPreferences(
                getString(R.string.preference_isLoggedIn),
                Context.MODE_PRIVATE
        );

        // Jika sudah memiliki shared preferences, langsung masuk.
        if (sharedPref.getBoolean(getString(R.string.preference_isLoggedIn), false)) {
            openMainMenu();
        }

        Button btnLupaPassword, btnLogin;
        EditText loginText, passwordText;

        btnLupaPassword = findViewById(R.id.lupapassword);
        loginText = findViewById(R.id.username);
        passwordText = findViewById(R.id.pass);
        btnLogin = findViewById(R.id.Login);

        btnLupaPassword.setOnClickListener((v) -> openLupaPassword());

        btnLogin.setOnClickListener((v) -> {
            String username = loginText.getText().toString();
            String password = passwordText.getText().toString();

            // Ketika seorang user berhasil login, setting 'sharedPreference' menjadi true.
            // Hal ini mencegah login ulang ketika seseorang mengakses aplikasi lagi.
            // Segi security, aman karena hanya menyimpan boolean variable.
            if (username.equals("user") && (password.equals("123"))) {
                SharedPreferences.Editor editor = sharedPref.edit();
                editor.putBoolean(getString(R.string.preference_isLoggedIn), true);
                editor.apply();

                openMainMenu();
            } else {
                Toast.makeText(MainActivity.this, "Username / Password Salah.", Toast.LENGTH_SHORT).show();
            }
        });

        // Manajemen permissions ada disini.
        if (ContextCompat.checkSelfPermission(getApplicationContext(), Manifest.permission.CAMERA)
                == PackageManager.PERMISSION_DENIED) {
            if (ActivityCompat.shouldShowRequestPermissionRationale(this, Manifest.permission.CAMERA)) {
                ActivityCompat.requestPermissions(
                        this,
                        new String[]{Manifest.permission.CAMERA},
                        cameraRequestCode
                );
            } else {
                ActivityCompat.requestPermissions(
                        MainActivity.this,
                        new String[]{Manifest.permission.CAMERA},
                        cameraRequestCode
                );
            }
        }
    }

    public void openLupaPassword() {
        Intent intent = new Intent(MainActivity.this, LupaPassword.class);
        startActivity(intent);
    }

    public void openMainMenu() {
        Intent intent = new Intent(MainActivity.this, MainMenu.class);
        startActivity(intent);
    }
}

File: app/src/main/java/avanger/co/id/MainActivity.kt
package avanger.co.id

import android.content.Intent
import android.os.Bundle
import android.view.View
import androidx.appcompat.app.AppCompatActivity
import avanger.co.id.databinding.ActivityMainBinding
import com.google.android.material.snackbar.Snackbar
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.auth.ktx.auth
import com.google.firebase.ktx.Firebase

class MainActivity : AppCompatActivity() {
    private lateinit var auth: FirebaseAuth
    private lateinit var binding: ActivityMainBinding

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)

        // Firebase Authentication
        auth = Firebase.auth

        binding.lupapassword.setOnClickListener { openLupaPassword() }

        binding.login.setOnClickListener {
            val username = binding.username.text.toString()
            val password = binding.password.text.toString()
            val usernameSuffix = getString(R.string.firebase_email)
            val firebaseUsername = "$username@$usernameSuffix"

            // Tampilkan progress bar.
            binding.formProgress.visibility = View.VISIBLE

            // Ketika seorang user berhasil login, tidak perlu apa-apa lagi karena sudah berada di Firebase.
            // Segi security, sangat aman - Firebase does all of the abstractions.
            auth.signInWithEmailAndPassword(firebaseUsername, password).addOnCompleteListener { task ->
                binding.formProgress.visibility = View.GONE

                if (task.isSuccessful) {
                    openMainMenu()
                } else {
                    Snackbar.make(binding.activityMainView, getString(R.string.main_fail_auth), Snackbar.LENGTH_LONG).show()
                }
            }
        }
    }

    private fun openLupaPassword() {
        val intent = Intent(this@MainActivity, LupaPassword::class.java)
        startActivity(intent)
    }

    private fun openMainMenu() {
        val intent = Intent(this@MainActivity, MainMenu::class.java)
        startActivity(intent)
        finish()
    }
}